### PR TITLE
Epoch change processing fixes

### PR DESCRIPTION
**Pull request by sasurobert**

### CODE REVIEW

File: process/block/shardblock_test.go [hunk 1]
```diff
@@ -4281,3 +4281,173 @@ func TestShardProcessor_updateStateStorage(t *testing.T) {
...
+
+	blockChain := &mock.BlockChainMock{GetCurrentBlockHeaderCalled: func() data.HeaderHandler {
+		return &block.Header{Epoch: epochStartTrigger.Epoch() - 1, Round: epochStartTrigger.EpochFinalityAttestingRound() + process.EpochChangeGracePeriod + 1}
+	}}
+
+	err := sp.CheckEpochCorrectnessCrossChain(blockChain)
+	assert.Equal(t, process.ErrMissingHeader, err)
+}
+
+func TestShardProcessor_checkEpochCorrectnessCrossChainInCorrectEpochRollback1(t *testing.T) {
```
Comment (SebastianMarian) [inline process/block/shardblock_test.go]: Rollback1Block

File: process/block/shardblock.go [hunk 2]
```diff
@@ -680,6 +721,12 @@ func (sp *shardProcessor) CommitBlock(
 		return err
 	}
 
+	err = sp.checkEpochCorrectnessCrossChain(chainHandler)
```
Comment (SebastianMarian) [inline process/block/shardblock.go]: This code should be executed now without the initial condition if header.IsStartOfEpochBlock() {  ?

File: process/block/shardblock.go [hunk 3]
```diff
@@ -680,6 +721,12 @@ func (sp *shardProcessor) CommitBlock(
...
+	sp.epochStartTrigger.SetProcessed(header)
```
Comment (SebastianMarian) [inline process/block/shardblock.go]: It is ok to set processed this header even if there is an err return by checkEpochCorrectnessCrossChain ?

File: process/block/shardblock.go [hunk 4]
```diff
@@ -258,6 +258,14 @@ func (sp *shardProcessor) ProcessBlock(
 		return err
 	}
 
+	if header.IsStartOfEpochBlock() {
+		err = sp.checkEpochCorrectnessCrossChain(chainHandler)
+		sp.epochStartTrigger.SetProcessed(header)
```
Comment (SebastianMarian) [inline process/block/shardblock.go]: It is ok to set processed this header even if there is an err return by checkEpochCorrectnessCrossChain ?

File: process/block/shardblock.go [hunk 5]
```diff
@@ -258,6 +258,14 @@ func (sp *shardProcessor) ProcessBlock(
 		return err
 	}
 
+	if header.IsStartOfEpochBlock() {
```
Comment (SebastianMarian) [inline process/block/shardblock.go]: Do you need this part of code also in commit block again?

File: process/block/shardblock.go [hunk 6]
```diff
@@ -347,28 +355,62 @@ func (sp *shardProcessor) checkEpochCorrectness(
...
+		}
+
+		metaBlockHash, err := core.CalculateHash(sp.marshalizer, sp.hasher, metaBlock)
+		if err != nil {
+			return fmt.Errorf("%w could not calculate hash for epoch start metablock for epoch %d",
+				err, header.GetEpoch())
+		}
+
+		if !bytes.Equal(header.EpochStartMetaHash, metaBlockHash) {
+			log.Warn("epoch start meta hash missmatch", "proposed", header.EpochStartMetaHash, "calculated", sp.epochStartTrigger.EpochStartMetaHdrHash())
```
Comment (SebastianMarian) [inline process/block/shardblock.go]: metaBlockHash instead sp.epochStartTrigger.EpochStartMetaHdrHash()

File: process/block/shardblock_test.go (hunk 1 above)
Comment (sasurobert) [reply process/block/shardblock_test.go]: done.

File: process/block/shardblock.go (hunk 2 above)
Comment (sasurobert) [reply process/block/shardblock.go]: initial condition is not more needed. It is better to check every time.

File: process/block/shardblock.go (hunk 3 above)
Comment (sasurobert) [reply process/block/shardblock.go]: put set only if not error.

File: process/block/shardblock.go (hunk 4 above)
Comment (sasurobert) [reply process/block/shardblock.go]: changed. put the verification only if isstartofepochblock and setprocessed only if epochcorrectess does not return error.

File: process/block/shardblock.go (hunk 5 above)
Comment (sasurobert) [reply process/block/shardblock.go]: yes. As the leader cannot make revert in create or applybodytoheader.

File: process/block/shardblock.go (hunk 6 above)
Comment (sasurobert) [reply process/block/shardblock.go]: done.

File: process/block/shardblock.go (hunk 5 above)
Comment (SebastianMarian) [reply process/block/shardblock.go]: ok

File: process/block/shardblock.go [hunk 7]
```diff
@@ -262,6 +262,13 @@ func (sp *shardProcessor) ProcessBlock(
 		return err
 	}
 
+	if header.IsStartOfEpochBlock() {
+		err = sp.checkEpochCorrectnessCrossChain(chainHandler)
+		if err != nil {
+			return err
+		}
+	}
```
Comment (SebastianMarian) [inline process/block/shardblock.go]: Do you forgot to add this line after checking the err? sp.epochStartTrigger.SetProcessed(header)

**Pull request merged**